README
Future-Q

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Future::Q

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Future-Q

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Future-Q

    CPAN Ratings
        http://cpanratings.perl.org/d/Future-Q

    Search CPAN
        http://search.cpan.org/dist/Future-Q/


LICENSE AND COPYRIGHT

Copyright (C) 2013 Toshio Ito

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Future-Q

0.100    2014-12-07
         [BUG FIX]
         - It did not call SUPER::DESTROY when it's DESTROYed.
           This was not a big deal unless you used the debug mode of Future.
         - is_rejected() method now returns a boolified value.
         - Test t/10-construction.t emitted warning messages with Future 0.30.
           The offending tests are removed.
         [ENHANCEMENT]
         - Slight performance enhancements of want_all() etc. (5 - 15% boost)
         [PACKAGING]
         - Migrate to Module::Build::Prereqs::FromCPANfile.

0.090    2014-03-30
         [API CHANGE]
         - Add finally() method.
         [DEPENDECY UPDATE]
         - Add dependency on parent pragma.
         [DOCUMENTATION UPDATE]
         - Mention AnyEvent::Promise.

0.080    2014-03-24
         [DEPENDENCY UPDATE]
         - Now it requires Future 0.22 to pass the tests.

0.070    2014-03-24
         [API CHANGE]
         - Add $OnError package variable.

0.060    2014-03-23
         [API CHANGE]
         - Add resolve() method.
         [DOCUMENTATION UPDATE]
         - Promises module no longer emulates jQuery.Deferred. Now it's more like Q.

0.050    2013-11-17
         - No changes to the code.
         - Documentation update. Some interesting modules are added to "SEE ALSO" section.
         - Switch to Module::Build::Pluggable with cpanfile.

0.040    2013-07-15
         - [BUG FIX] Now Carp::croak() called in callback functions for try(), fcall(),
           then() and catch() methods report the error from the user's perspective.
           Note that you have to include 'Future::Q' in @CARP_NOT of the package
           that calls try(), fcall(), then() or catch().
         - Perhaps error reporting might need a little more brush-up...

0.030    2013-07-03
         - No changes to the code.
         - Update POD in response to Future 0.14 and q.js v0.9.5
         - Confirmed it passes the tests with Future 0.15 installed.

0.020    2013-05-12
         - Support Future 0.13
           POD is updated now that Future has its own then() method.
         - [BUG FIX] It crashed when a Future object (not a Future::Q object) was returned
           from then() callbacks. This bug is fixed.
         - Drop support for Future::Utils::repeat() in the test suite. repeat() may not
           go well with Future::Q's feature of reporting unhandled failures.
         - Now Module::Build is used for packaging.

0.012    2013-04-14
         Initial release.



t/testlib/Utils.pm
package testlib::Utils;
use strict;
use warnings;
use Exporter qw(import);
use Future::Q;
use Test::Builder;
use Test::More;
use Scalar::Util qw(refaddr);
use Carp;

our @EXPORT_OK = qw(newf init_warn_handler test_log_num filter_callbacks is_immediate isnt_identical);

sub newf {
    return Future::Q->new;
}

sub init_warn_handler {
    delete $ENV{PERL_FUTURE_DEBUG};
    $SIG{__WARN__} = sub {
        confess @_;  ## no warning should be printed
    };
}

sub test_log_num {
    my ($testee_code, $exp_log_num, $msg) = @_;
    local $Test::Builder::Level = $Test::Builder::Level + 1;
    my @logs = ();
    local $Future::Q::OnError = sub { push @logs, shift };
    {
        local $Test::Builder::Level = $Test::Builder::Level + 1;
        $testee_code->();
    }
    is(int(@logs), $exp_log_num, $msg) or diag(explain @logs);
}

sub filter_callbacks {
    my ($case_arg, $on_done, $on_fail) = @_;
    my %switch = (
        on_done => sub { ($on_done) },
        on_fail => sub { (undef, $on_fail) },
        both    => sub { ($on_done, $on_fail) },
        none    => sub { () },
    );
    return $switch{$case_arg}->();
}

sub is_immediate {
    my ($case_string) = @_;
    my %switch = (
        immediate_done => sub { 1 },
        immediate_fail => sub { 1 },
        immediate_cancel => sub { 1 },
        pending_done => sub { 0 },
        pending_fail => sub { 0 },
        pending_cancel => sub { 0 },
    );
    return $switch{$case_string}->();
}

sub isnt_identical {
    my ($got, $not_exp, $msg) = @_;
    local $Test::Builder::Level = $Test::Builder::Level + 1;
    isnt(refaddr($got), refaddr($not_exp), $msg);
}

1;


Build.PL
use 5.006;
use strict;
use warnings;
use Module::Build::Pluggable (
    'CPANfile'
);
 
my $builder = Module::Build::Pluggable->new(
    module_name         => 'Future::Q',
    license             => 'perl',
    dist_author         => q{Toshio Ito <toshioito@cpan.org>},
    dist_version_from   => 'lib/Future/Q.pm',
    release_status      => 'stable',
    add_to_cleanup     => [ 'Future-Q-*' ],
    recursive_test_files => 1,
    no_index => {
        directory => ["t", "xt", "eg", "inc"],
        file => ['README.pod'],
    },
    meta_add => {
        'meta-spec' => {
            version => 2,
            url => 'https://metacpan.org/pod/CPAN::Meta::Spec',
        },
        resources => {
            bugtracker => {
                web => 'https://github.com/debug-ito/Future-Q/issues',
            },
            repository => {
                url => 'git://github.com/debug-ito/Future-Q.git',
                web => 'https://github.com/debug-ito/Future-Q',
            },
        }
    }
);
 
$builder->create_build_script();


q/fin.js

var q = require("q");
var async = require("async");

var task_queue = async.queue(function(task, done) {
    console.log("----- " + task.label);
    var d = q.defer();
    task.code(done, d);
}, 1);

function test(label, code) {
    task_queue.push({label: label, code: code});
}

function finish(done, promise) {
    return promise.then(function(r) {
        console.log("Fulfilled: " + r);
        done();
    }, function(e) {
        console.log("Rejected: " + e);
        done();
    });
}

function print_args(args) {
    var i;
    if(args.length === 0) {
        console.log("No arg");
        return;
    }
    for(i = 0 ; i < args.length ; i++) {
        console.log("arg["+i+"]" + args[i]);
    }
}

function explain(name, promise) {
    return name + " is " + promise.inspect().state;
}

test("done -> normal", function(done, d) {
    finish(done, d.promise.finally(function(hoge) {
        print_args(arguments);
        return "return";
    }));
    d.resolve("orig");
});

test("done -> die", function(done, d) {
    finish(done, d.promise.finally(function() {
        print_args(arguments);
        throw "return!";
    }));
    d.resolve("orig");
});

test("done -> done promise", function(done, d) {
    var dd = q.defer();
    var next = d.promise.finally(function() {
        print_args(arguments);
        console.log("d is done. " + explain("next", next));
        q.delay(100).then(function() {
            console.log("timeout. " + explain("next", next));
            dd.resolve("return");
        });
        return dd.promise;
    });
    finish(done, next);
    d.resolve("orig");
});

test("done -> fail promise", function(done, d) {
    var dd = q.defer();
    var next = d.promise.finally(function() {
        print_args(arguments);
        console.log("d is done. " + explain("next", next));
        q.delay(100).then(function() {
            console.log("timeout. " + explain("next", next));
            dd.reject("return!");
        });
        return dd.promise;
    });
    finish(done, next);
    d.resolve("done");
});

test("done -> no_callback", function(done, d) {
    finish(done, d.promise.finally());
    d.resolve("orig");
});

test("fail -> normal", function(done, d) {
    finish(done, d.promise.finally(function() {
        print_args(arguments);
        return "return";
    }));
    d.reject("orig!");
});

test("fail -> die", function(done, d) {
    finish(done, d.promise.finally(function() {
        print_args(arguments);
        throw "return!";
    }))
    d.reject("orig!");
})

test("fail -> done promise", function(done, d) {
    var dd = q.defer();
    var next = d.promise.finally(function() {
        print_args(arguments);
        console.log("d is failed. " + explain("next", next));
        q.delay(100).then(function() {
            console.log("timeout. " + explain("next", next));
            dd.resolve("return");
        });
        return dd.promise;
    });
    finish(done, next);
    d.reject("orig!");
});

test("fail -> fail promise", function(done, d) {
    var dd = q.defer();
    var next = d.promise.finally(function() {
        print_args(arguments);
        console.log("d is failed. " + explain("next", next));
        q.delay(100).then(function() {
            console.log("timeout. " + explain("next", next));
            dd.reject("return!");
        });
        return dd.promise;
    });
    finish(done, next);
    d.reject("orig!");
});

test("fail -> no_callback", function(done, d) {
    finish(done, d.promise.finally());
    d.reject("orig!");
});



q/resolve.js

var q = require("q");
var async = require("async");

var show_state = function(d, label) {
    var prefix = label ? label + " " : "Result ";
    var content = d.promise.inspect();
    console.log(prefix + "state: " + content.state);
    if(content.state === "fulfilled") {
        console.log(prefix + "value: " + content.value);
    }else if(content.state === "rejected") {
        console.log(prefix + "reason: " + content.reason);
    }

};

var test_queue = async.queue(function(task, done) {
    var label = task.label;
    var code = task.code;
    var d = q.defer();
    console.log("---- " + label);
    code(d, function() {
        show_state(d);
        done();
    });
}, 1);

var test = function(label, code) {
    test_queue.push({label: label, code: code});
};

test("normal", function(d, done) {
    d.resolve("AAA");
    done();
});

test("fulfilled promise", function(d, done) {
    var dd = q.defer();
    dd.resolve("BBB");
    
    dd.promise.then(function() {
        console.log("> given promise state: " + dd.promise.inspect().state);
        d.resolve(dd.promise);
        done();
    });
});

test("rejected promise", function(d, done) {
    var dd = q.defer();
    dd.reject("CCC");
    dd.promise.then(null, function() {
        console.log("> given promise state: " + dd.promise.inspect().state);
        d.resolve(dd.promise);
        done();
    });
});

test("pending fulfilled promise", function(d, done) {
    var dd = q.defer();
    d.resolve(dd.promise);
    show_state(d, "d (dd pending):");
    dd.resolve("DDD");
    show_state(dd, "dd:");
    done();
});

test("pending rejected promise", function(d, done) {
    var dd = q.defer();
    d.resolve(dd.promise);
    show_state(d, "d (dd pending):");
    dd.reject("EEE");
    show_state(dd, "dd:");
    done();
});

test("pending fulfilled, try to reject while pending", function(d, done) {
    var dd = q.defer();
    d.resolve(dd.promise);
    show_state(d, "d (dd pending):");
    d.reject("HOGEHOGE");
    show_state(d, "d (tried to reject):");
    dd.resolve("FFF");
    done();
});





eg/synopsis.pl
use strict;
use warnings;

sub other_async_func {
    my (%args) = @_;
    warn "--- other_async_func\n";
    @_ = ("failure");
    goto $args{on_failure};
}

sub bad_func {
    die "something terrible happened.";
}


{
    use Future::Strict;

    sub async_func_future {
        my $f = Future::Strict->new;
        ## Assume other_async_func() always fails
        other_async_func(
            on_success => sub { $f->done(@_) },
            on_failure => sub { $f->fail(@_) },
        );
        return $f;
    }

    {
        ### CASE 1: It complains
        async_func_future();
    }

    {
        ### CASE 2: It complains
        async_func_future()->on_done(sub {
            my $result = shift;
            print "OK: $result\n";
        });
    }

    {
        ### CASE 3: It does NOT complain
        async_func_future()->on_done(sub {
            my $result = shift;
            print "OK: $result\n";
        })->on_fail(sub {
            my $failure = shift;
            print "NG: $failure\n";
        });
    }
}


{
    {
        ### CASE 4: It complains (if bad_func() throws an exception)
        my $f_result = Future::Strict->new->done("start")->and_then(sub {
            my $f = shift;
            my $result = bad_func($f->get);
            return Future::Strict->new->done($result);
        });
    }
}



lib/Future/Strict.pm
package Future::Strict;
use strict;
use warnings;
use Future 0.12;
use base "Future";
use Devel::GlobalDestruction;
use Scalar::Util qw(refaddr blessed);
use Carp;
use Try::Tiny;

## ** lexical attributes to avoid collision of names.

my %failure_handled_for = ();

sub new {
    my ($class, @args) = @_;
    my $self = $class->SUPER::new(@args);
    my $id = refaddr $self;
    $failure_handled_for{$id} = 0;
    return $self;
}

sub DESTROY {
    my ($self) = @_;
    return if in_global_destruction;
    my $id = refaddr $self;
    if($self->is_ready && $self->failure && !$failure_handled_for{$id}) {
        $self->_warn_failure();
        my @failed_subfutures = try {
            $self->failed_futures;
        }catch {
            ();
        };
        foreach my $f (@failed_subfutures) {
            $f->_warn_failure(is_subfuture => 1) if blessed($f) && $f->can('_warn_failure');
        }
    }
    delete $failure_handled_for{$id};
}

sub _set_failure_handled {
    my ($self) = @_;
    $failure_handled_for{refaddr $self} = 1;
}

sub _warn_failure {
    my ($self, %options) = @_;
    if($self->is_ready && $self->failure) {
        my $failure = $self->failure;
        if($options{is_subfuture}) {
            carp "Failure of subfuture $self may not be handled: $failure  subfuture may be lost";
        }else {
            carp "Failure of $self is not handled: $failure  future is lost";
        }
    }
}

sub on_fail {
    my ($self) = @_;
    $self->_set_failure_handled();
    goto $self->can('SUPER::on_fail');
}

sub on_ready {
    my ($self) = @_;
    $self->_set_failure_handled();
    goto $self->can('SUPER::on_ready');
}

foreach my $method (qw(wait_all wait_any needs_all needs_any)) {
    no strict "refs";
    my $supermethod = "SUPER::$method";
    *{$method} = sub {
        my ($self, @subfutures) = @_;
        foreach my $sub (@subfutures) {
            next if !blessed($sub) || !$sub->can('_set_failure_handled');
            $sub->_set_failure_handled();
        }
        goto $self->can($supermethod);
    };
}


=head1 NAME

Future::Strict - a strict future that will complain when it fails and is not handled.

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';


=head1 SYNOPSIS

    use Future::Strict;
    
    sub async_func_future {
        my $f = Future::Strict->new;
        ## Assume other_async_func() always fails
        other_async_func(
            on_success => sub { $f->done(@_) },
            on_failure => sub { $f->fail(@_) },
        );
        return $f;
    }

    {
        ### CASE 1: It complains
        async_func_future();
    }

    {
        ### CASE 2: It complains
        async_func_future()->on_done(sub {
            my $result = shift;
            print "OK: $result\n";
        });
    }

    {
        ### CASE 3: It does NOT complain
        async_func_future()->on_done(sub {
            my $result = shift;
            print "OK: $result\n";
        })->on_fail(sub {
            my $failure = shift;
            print "NG: $failure\n";
        });
    }

=head1 DESCRIPTION

L<Future::Strict> is a subclass of L<Future>.
It extends the original L<Future> so that it warns you
when a L<Future::Strict> object in the failure state is
destroyed but its failure has never been handled.


=head2 What's the benefit of Future::Strict?

The benefit of using L<Future::Strict> instead of regular L<Future>
is that it can detect the possibly dangerous situation when
a future fails but its failure is never handled.

In the L</SYNOPSIS>, the C<async_func_future()> function returns
a future that may either succeed or fail. B<Let's assume it always fails in this example>.

However, with regular L<Future> it is very easy to ignore failures
if you are not very careful.

For example, if you are interested only in the side-effect of C<async_func_future()> but
not in its result, you are very likely to write CASE 1, that is, just throwing away the
returned future. Or if you are too lazy to set the failure handler, you will probably
write CASE 2. In both cases, the returned failure is discarded.

If this happens with L<Future::Strict>, it prints warning message
to motivate you to handle the failures properly.

L<Future::Strict> is even more beneficial when you use chaining methods such as C<and_then()>.
This is because as of L<Future> 0.11 B<< exceptions thrown in callbacks for C<and_then()>, C<or_else>
and C<followed_by()> are caught and transformed into failed futures. >>

For example, the following code seems to involve no failed future.

    {
        ### CASE 4: It complains (if bad_func() throws an exception)
        my $f_result = Future::Strict->new->done("start")->and_then(sub {
            my $f = shift;
            my $result = bad_func($f->get);
            return Future::Strict->new->done($result);
        });
    }

However, if C<bad_func()> throws an exception, it is silently transformed into a failed future.
As a result, C<$f_result> becomes a failed future.
If you just discard C<$f_result> like this example, the exception is never handled.
What's worse, if you don't use L<Future::Strict>, the exception is never visible to you,
which can lead to very hard-to-track bugs.

L<Future::Strict> makes failed futures visible to you.
With L<Future::Strict> you will not miss unexpected failed futures in most cases.


=head2 When and how does a Future::Strict complain?

A failed L<Future::Strict> object prints warning messages when it is destroyed.

The warning messages are printed through Perl's warning facility.
You can capture them by setting C<< $SIG{__WARN__}. >>
The warning messages can be evaluated to strings.
(They ARE strings actually, but this may change in future versions)


=head2 How can I convince a Future::Strict that its failure is handled?

To prevent a failed L<Future::Strict> from complaining,
you have to convince it that its failure is handled before it's destroyed.

L<Future::Strict> thinks failures of the following futures are handled.

=over

=item *

Futures that C<on_fail()> or C<on_ready()> method is called on.

=item *

Futures that C<and_then()>, C<or_else()> or C<followed_by()> method is called on.

=item *

Futures returned by the callbacks for C<and_then()>, C<or_else()> or C<followed_by()> method.

=item *

Subfutures given to C<wait_all()>, C<wait_any()>, C<needs_all()> or C<needs_any()> method.

=back

Therefore, remember to call C<on_fail()> on any future that may fail and
handle the failure in its callback.


=head1 CAVEAT

If you don't want to miss failed futures, I recommend you to follow the guidelines below.

=over

=item *

Do not use C<on_ready()> or C<followed_by()> method unless it's absolutely necessary.
In callbacks for these methods you may forget to handle failures,
but L<Future::Strict> thinks they are handled.


=item *

Always inspect failed subfutures by C<failed_futures()> method
in callbacks for dependent futures returned by C<wait_all()>, C<wait_any()>,
C<needs_all()> and C<needs_any()>.

This is because there may be multiple of failed subfutures.
It is even possible that some subfutures fail but the dependent future succeeds.

=back


=head1 METHODS

L<Future::Strict> inherits all the class and object methods from L<Future>.
There is no extra public method.

=head1 SEE ALSO

L<Future>

=head1 ACKNOWLEDGEMENT

Paul Evans, C<< <leonerd at leonerd.org.uk> >> - author of L<Future>


=head1 AUTHOR

Toshio Ito, C<< <toshioito at cpan.org> >>


=head1 LICENSE AND COPYRIGHT

Copyright 2013 Toshio Ito.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut


1; # End of Future::Strict


Makefile.PL
use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME                => 'Future::Q',
    AUTHOR              => q{Toshio Ito <toshioito@cpan.org>},
    VERSION_FROM        => 'lib/Future/Q.pm',
    ABSTRACT_FROM       => 'lib/Future/Q.pm',
    ($ExtUtils::MakeMaker::VERSION >= 6.3002
      ? ('LICENSE'=> 'perl')
      : ()),
    PL_FILES            => {},
    PREREQ_PM => {
        'Future' => '0.13',
        'Devel::GlobalDestruction' => 0,
        'Scalar::Util' => 0,
        'Carp' => 0,
        'Try::Tiny' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More' => 0,
        'FindBin' => 0,
        'Carp' => 0,
        'Scalar::Util' => 0,
        'Test::MockModule' => "0.05",
        'Test::Identity' => 0,
        'Test::Memory::Cycle' => 0,
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'Future-Q-*' },
);
